src/passes/PostAssemblyScript.cpp

```diff
@@ -0,0 +1,530 @@
+  void computeInfluences() {
+    for (auto& pair : locations) {
+      auto* curr = pair.first;
+      if (auto* set = curr->dynCast<LocalSet>()) {
+        if (auto* get = set->value->dynCast<LocalGet>()) {
+          getInfluences[get].insert(set);
+        }
+      } else {
+        auto* get = curr->cast<LocalGet>();
+        for (auto* set : getSetses[get]) {
+          setInfluences[set].insert(get);
+        }
+      }
+    }
+  }
+};
+
+// A pass that eliminates redundant retain and release calls.
+//
+// Does a cheap traversal first, remembering ARC-style patterns, and goes all-in
+// only if it finds any.
+//
+// This is based on the assumption that the compiler is not allowed to emit
+// unbalanced retains or releases, except if a value is returned or otherwise
+// escapes in one branch. In turn, we do not have to deal with control
+// structures but can instead look for escapes reached (by any alias) using
+// AliasGraph.
+//
+// For example, in code like
+//
+//   var a = __retain(X);
+//   if (cond) {
+//     return a;
+//   }
+//   __release(a);
+//  return null;
+//
+// we cannot eliminate the retain/release pair because the implementation
+// dictates that returned references must remain retained for the caller since
+// dropping to RC=0 on the boundary would prematurely free the object.
+//
+// Typical patterns this recognizes are
+//
+//   var a = __retain(X);
+//   __release(a); // simple pair
+//
+//   var a = __retain(X);
+//   if (cond) {
+//     __release(a); // non-escaping...
+//   } else {
+//     __release(a); // ...balanced pair
+//   }
+//
+//   var a;
+//   if (cond) {
+//     a = __retain(X);
+//   } else {
+//     a = __retain(Y);
+//   }
+//   __release(a); // balanced retain
+//
+// including technically invalid patterns assumed to be not present in compiler
+// output, like:
+//
+//   var b = __retain(a);
+//   if (cond) {
+//     __release(b); // unbalanced release
+//   }
+//
+// To detect the latter, we'd have to follow control structures around, which
+// we don't do since it isn't neccessary / to keep the amount of work minimal.
+struct OptimizeARC : public WalkerPass<PostWalker<OptimizeARC>> {
+
+#ifndef POST_ASSEMBLYSCRIPT_DEBUG
+  bool isFunctionParallel() override { return true; }
+#endif
+
+  Pass* create() override { return new OptimizeARC; }
+
+  // Sets that are retains, to location
+  std::unordered_map<LocalSet*, Expression**> retains;
+
+  // Gets that are releases, to location
+  std::unordered_map<LocalGet*, Expression**> releases;
+
+  // Gets that are escapes
+  std::unordered_set<LocalGet*> escapes;
+
+  void visitLocalSet(LocalSet* curr) {
+    // local.set(X, __retain(...)) ?
+    if (auto* call = curr->value->dynCast<Call>()) {
+      if (call->target == RETAIN) {
+        retains[curr] = getCurrentPointer();
+      }
+    }
+  }
+
+  void visitCall(Call* curr) {
+    // __release(local.get(X, ...)) ?
+    if (curr->target == RELEASE && curr->operands.size() == 1) {
```

Comment:
(or with an `isRelease` this would be even shorter)

Reply:
Yep, added `isRetain` etc. helpers. There are multiple of these because parts of the passes look at slightly different patterns, like the first pass wanting full retain patterns including a `local.set` while the second just looks for plain calls. There's also no `isRelease` since we can't tell from the `local.get` (which we use as the key) in `release(local.get(...))` alone that it is a release.